dex/pair/src/safe_price_views.rs

```diff
@@ -0,0 +1,287 @@
+            &price_observations,
+            search_round,
+        )
+    }
+
+    fn compute_weighted_price(
+        &self,
+        input_payment: EsdtTokenPayment<Self::Api>,
+        first_price_observation: &PriceObservation<Self::Api>,
+        last_price_observation: &PriceObservation<Self::Api>,
+    ) -> EsdtTokenPayment<Self::Api> {
+        let first_token_id = self.first_token_id().get();
+        let second_token_id = self.second_token_id().get();
+
+        require!(
+            input_payment.token_identifier == first_token_id
+                || input_payment.token_identifier == second_token_id,
+            ERROR_BAD_INPUT_TOKEN
+        );
+
+        let (weighted_first_token_reserve, weighted_second_token_reserve) =
+            self.compute_weighted_reserves(first_price_observation, last_price_observation);
+
+        let output_payment = if input_payment.token_identifier == first_token_id {
+            let output_amount =
+                input_payment.amount * weighted_second_token_reserve / weighted_first_token_reserve;
+            EsdtTokenPayment::new(second_token_id, 0, output_amount)
+        } else {
+            let output_amount =
+                input_payment.amount * weighted_first_token_reserve / weighted_second_token_reserve;
+            EsdtTokenPayment::new(first_token_id, 0, output_amount)
+        };
+        require!(output_payment.amount > 0u64, ERROR_ZERO_AMOUNT);
+        output_payment
+    }
+
+    fn get_price_observation(
+        &self,
+        current_index: usize,
+        max_observations: usize,
+        price_observations: &ManagedVec<PriceObservation<Self::Api>>,
+        search_round: Round,
+    ) -> PriceObservation<Self::Api> {
+        if price_observations.is_empty() {
+            sc_panic!(ERROR_SAFE_PRICE_OBSERVATION_DOES_NOT_EXIST)
+        }
+
+        // Check if the requested price observation is the last one
+        let last_observation = price_observations.get(current_index);
+        if last_observation.recording_round <= search_round {
+            return last_observation;
+        }
+
+        // Check if the observation round exists in the list
+        let oldest_observation_index = if price_observations.len() < max_observations {
+            0
+        } else {
+            (current_index + 1) % max_observations
+        };
+        let oldest_observation_option = price_observations.try_get(oldest_observation_index);
+        match oldest_observation_option {
+            Some(oldest_observation) => {
+                if search_round == oldest_observation.recording_round {
+                    return oldest_observation;
+                }
+                require!(
+                    search_round > oldest_observation.recording_round,
+                    ERROR_SAFE_PRICE_OBSERVATION_DOES_NOT_EXIST
+                );
+            }
+            None => sc_panic!(ERROR_SAFE_PRICE_OBSERVATION_DOES_NOT_EXIST),
+        }
+
+        // Binary search algorithm
```

Comment:
function too big - please separate in 3 smaller functions - binary search has to be one.

Reply:
Updated.